src/vterm.c

```diff
@@ -0,0 +1,474 @@
+
+  if (buffer == NULL)
+    {
+      *length = required_buf_size;
+      return true;
+    }
+
+  if (*length < required_buf_size)
+    {
+      *length = required_buf_size;
+      xsignal0 (Qargs_out_of_range);
+    }
+
+  *length = required_buf_size;
+  memcpy (buffer, SDATA (lisp_str_utf8), raw_size + 1);
+
+  return true;
+}
+
+int
+vterm_string_bytes(Lisp_Object string) {
+  ptrdiff_t size = 0;
+  vterm_module_copy_string_contents(string, NULL, &size);
+  return size;
+}
+
+void byte_to_hex(uint8_t byte, char *hex) { snprintf(hex, 3, "%.2X", byte); }
+
+Lisp_Object
+color_to_rgb_string(VTermColor color) {
+  char buffer[8];
+  buffer[0] = '#';
+  buffer[7] = '\0';
+  byte_to_hex(color.red, buffer + 1);
+  byte_to_hex(color.green, buffer + 3);
+  byte_to_hex(color.blue, buffer + 5);
+
+  /* return env->make_string(env, buffer, 7); */
+  return make_string (buffer, 7);
+}
+
+uint8_t hex_to_byte(char *hex) { return strtoul(hex, NULL, 16); }
+
+VTermColor
+rgb_string_to_color(Lisp_Object string) {
+  VTermColor color;
+  ptrdiff_t len = 8;
+  char buffer[len];
+  char hex[3];
+  
+  vterm_module_copy_string_contents (string, buffer, &len);
+
+  hex[0] = buffer[1];
+  hex[1] = buffer[2];
+  color.red = hex_to_byte(hex);
+  hex[0] = buffer[3];
+  hex[1] = buffer[4];
+  color.green = hex_to_byte(hex);
+  hex[0] = buffer[5];
+  hex[1] = buffer[6];
+  color.blue = hex_to_byte(hex);
+
+  return color;
+};
+
+void term_process_key(Term *term, unsigned char *key, size_t len,
+                             VTermModifier modifier) {
+  if (is_key(key, len, "<return>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_ENTER, modifier);
+  } else if (is_key(key, len, "<tab>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_TAB, modifier);
+  } else if (is_key(key, len, "<backspace>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_BACKSPACE, modifier);
+  } else if (is_key(key, len, "<escape>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_ESCAPE, modifier);
+  } else if (is_key(key, len, "<up>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_UP, modifier);
+  } else if (is_key(key, len, "<down>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_DOWN, modifier);
+  } else if (is_key(key, len, "<left>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_LEFT, modifier);
+  } else if (is_key(key, len, "<right>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_RIGHT, modifier);
+  } else if (is_key(key, len, "<insert>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_INS, modifier);
+  } else if (is_key(key, len, "<delete>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_DEL, modifier);
+  } else if (is_key(key, len, "<home>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_HOME, modifier);
+  } else if (is_key(key, len, "<end>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_END, modifier);
+  } else if (is_key(key, len, "<prior>")) {
+    vterm_keyboard_key(term->vt, VTERM_KEY_PAGEUP, modifier);
+  } else if (is_key(key, len, "<f0>")) {
```

Comment:
Could maybe these be done parametrically by looking for prefix `<f` and suffix `>` then parsing the middle as being a decimal integer? That would extend beyond 12 then if required.

Reply:
Makes sense.